tools/run_tests/xds_k8s_test_driver/tests/failover_test.py

```diff
@@ -0,0 +1,128 @@
+# Copyright 2020 gRPC authors.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import logging
+from typing import List
+
+from absl import flags
+from absl.testing import absltest
+
+from framework import xds_k8s_testcase
+from framework.infrastructure import k8s
+from framework.test_app import server_app
+
+logger = logging.getLogger(__name__)
+flags.adopt_module_key_flags(xds_k8s_testcase)
+
+# Type aliases
+_XdsTestServer = xds_k8s_testcase.XdsTestServer
+_XdsTestClient = xds_k8s_testcase.XdsTestClient
+
+
+class FailoverTest(xds_k8s_testcase.RegularXdsKubernetesTestCase):
+    REPLICA_COUNT = 3
+    MAX_RATE_PER_ENDPOINT = 100
+
+    def setUp(self):
+        super().setUp()
+        self.secondary_server_runner = server_app.KubernetesServerRunner(
+            k8s.KubernetesNamespace(self.secondary_k8s_api_manager,
+                                    self.server_namespace),
+            deployment_name=self.server_name + '-alt',
```

Comment:
@sergiitk So this proved fraught. Originally I had the suffix here as `-alternate-region`, which didn't show any problems in any of the operations of the framework (as far as I could determine from the logs, but I could easily have missed something) but the deployment in this runner was never moving to ready. Eventually I was able to discover:

```
$ kubectl get svcneg -o yaml
     message: 'googleapi: Error 400: Invalid value for field ''networkEndpointGroup'':
        ''xds-k8s-security-server-20210729-0938-k5ad1-psm-grpc-server-alternate-region''.
        Must be a match of regex ''(?:[a-z](?:[-a-z0-9]{0,61}[a-z0-9])?)|[1-9][0-9]{0,19}'',
```

But this took me (admittedly very new to k8s) quite a while of digging around to even know to check this. I think I eventually noticed that there was no NEG for the deployment, and that led to the above kubectl check that revealed the error. But I couldn't identify a particular operation of the framework that actually failed, so it wasn't clear to me if/where we could put a check to help avoid future-me from running into this same too-long-of-a-name problem again and not knowing what the underlying problem is. Curious if you might have a better idea of what step actually led to this and were it could be checked for.

Reply:
Ooof, I see how this could've been frustrating. 

1. One hard-to-predict thing that changed with adding the "multiple framework runs at the same time" feature (#26542) is added resource suffix, with date-time and nonce, f.e. `-20210729-0938-k5ad1`. This is needed to make each run unique. However this added 20 chars to each resource name. And different resource types have different char limits. You can see in the error you're getting, there's a 61 char limit on NEG name: `{0,61}`. In many cases these char limits are implicit, and the only way to know they exist, is to just run the script, and check for errors related to the name field. That's the reason I tried to cut down on char limit, and replaced `-alternate` with `-alt` in all places I found.
2. The problem with debugging NEGs is that they are a [custom k8s extension](https://kubernetes.io/docs/concepts/extend-kubernetes/api-extension/custom-resources/) made by GCP for GCP. There's a bunch on opaque machinery invisible to k8s. That's also why it takes forever to delete a k8s namespace, when there are live GCP resources using a NEG associated with a service in the namespace (f.e. this happens when cleanup is done in incorrect order)
3. My recent life learning is that most cases I get stuck in debugging, it could've spared myself some time by asking someone who worked on that thing I'm debugging. I'm still pretty bad at applying this lesson though 😄